sklearn_numba_dpex/kmeans/drivers.py

```diff
@@ -190,12 +186,8 @@ def __call__(
             work_group_size=work_group_size,
         )
 
-        X_t = dpctl.tensor.from_numpy(
-            np.ascontiguousarray(X).T, device=self.device
-        )
-        centroids_t = dpctl.tensor.from_numpy(
-            np.ascontiguousarray(centers_init).T, device=self.device
-        )
```

Comment:
I noticed that when feeding to kernels numpy arrays (without conversion to dpctl.tensor first) that are not contiguous, the kernel will segfault.

It's likely that ascontiguousarray is indeed useless when using dpctl.tensor conversion because the conversion will trigger copy to an aligned array anyway but did you try it ?

Reply:
I am pretty sure `dpctl.tensor.from_numpy` will always allocate a contiguous shared USM buffer, whatever the contiguity pattern of the source array.

I see no reason it would ever attempt to do multiple non-contiguous allocations to create such a tensor with holes.